libvips/convolution/convi_hwy.cpp

```diff
@@ -0,0 +1,317 @@
+/* 20/08/22 kleisauke
+ * 	- initial implementation
+ * 07/09/22 kleisauke
+ * 	- implement using ReorderWidenMulAccumulate
+ * 29/11/22 kleisauke
+ * 	- prefer use of RearrangeToOddPlusEven
+ */
+
+/*
+
+	This file is part of VIPS.
+
+	VIPS is free software; you can redistribute it and/or modify
+	it under the terms of the GNU Lesser General Public License as published by
+	the Free Software Foundation; either version 2 of the License, or
+	(at your option) any later version.
+
+	This program is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU Lesser General Public License for more details.
+
+	You should have received a copy of the GNU Lesser General Public License
+	along with this program; if not, write to the Free Software
+	Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+	02110-1301  USA
+
+ */
+
+/*
+
+	These files are distributed with VIPS - http://www.vips.ecs.soton.ac.uk
+
+ */
+
+#ifdef HAVE_CONFIG_H
+#include <config.h>
+#endif /*HAVE_CONFIG_H*/
+#include <glib/gi18n-lib.h>
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <math.h>
+
+#include <vips/vips.h>
+#include <vips/vector.h>
+#include <vips/debug.h>
+#include <vips/internal.h>
+
+#include "pconvolution.h"
+
+#ifdef HAVE_HWY
+
+#undef HWY_TARGET_INCLUDE
+#define HWY_TARGET_INCLUDE "libvips/convolution/convi_hwy.cpp"
+#include <hwy/foreach_target.h>
+#include <hwy/highway.h>
+
+namespace HWY_NAMESPACE {
+
+using namespace hwy::HWY_NAMESPACE;
+
+using DI32 = ScalableTag<int32_t>;
+using DI16 = ScalableTag<int16_t>;
+using DU8 = ScalableTag<uint8_t>;
+constexpr DU8 du8;
+constexpr Rebind<uint8_t, DI16> du8x16;
+constexpr Rebind<uint8_t, DI32> du8x32;
+constexpr DI16 di16;
+constexpr DI32 di32;
+
+HWY_ATTR void
+vips_convi_uchar_hwy(VipsRegion *out_region, VipsRegion *ir, VipsRect *r,
+	int32_t ne, int32_t nnz, int32_t offset, const int32_t *HWY_RESTRICT offsets,
+	const int16_t *HWY_RESTRICT mant, int32_t exp)
+{
+#if HWY_TARGET != HWY_SCALAR
+	int32_t bo = VIPS_RECT_BOTTOM(r);
+
+#if HWY_ARCH_RVV || (HWY_ARCH_ARM_A64 && HWY_TARGET <= HWY_SVE)
+	/* Ensure we do not cross 128-bit block boundaries on RVV/SVE.
+	 */
+	const int32_t N = 16;
+#else
```

Comment:
Note to self: this conditional block can be removed once we use <code>Interleave**Whole**{Lower,Upper}</code> in favor of `Interleave{Lower,Upper}`. See e.g. commit: https://github.com/kleisauke/libvips/commit/d7cfe964217c8674d5674d81b2768700424baa5e.

Reply:
PR https://github.com/libvips/libvips/pull/3860 should fix this.